### DISCUSSION TITLE
Unable to get domain routes working

### DISCUSSION THREAD

**Discussion by bsdev90** [adonisjs/core#4598] [Help] [answered]
Hi,

For several months I've been developing an application using AdonisJS (v5). My application is divided into three parts (back, admin, pwa).
Until now I used http://127.0.0.1:3333/pwa to access my application.
Now that I have to put it into production, /back, /admin and /pwa will be replaced by https://pwa.myapp.com, http://back.myapp.com and https://admin.myapp.com.
In the documentation, it's written that you should use the `domain` parameter in the route.ts, which is what I've done (using myapp.local to test locally that it works):
```javascript
import Route from '@ioc:Adonis/Core/Route'

Route.group(() => {
  /* Login */
  Route.get('/login', './Front/SecurityController.login').as('auth.user.login') // Login
  Route.post('/login', './Front/SecurityController.doLogin')

  /* App */
  Route.group(() => {
    Route.get('/home', './Front/HomeController.index').as('home') // Home
    Route.get('/users/:slug', './Front/UsersController.show').as('users.show') // Users
  })
    .as('front')
    .middleware('auth:user')
}).domain('pwa.myapp.local')
```
This is a very simplified example of my route.ts file, but the tests were made with this version.

If I run the command `node ace list:routes` it gives me:
```bash
GET|HEAD    pwa.myapp.local/login ───── auth.user.login › ./Front/SecurityController.login
POST        pwa.myapp.local/login ───────────────────── ./Front/SecurityController.doLogin
GET|HEAD    pwa.myapp.local/home ─────────────── front.home › ./Front/HomeController.index
GET|HEAD    pwa.myapp.local/users/:slug ── front.users.show › ./Front/UsersController.show
```

I've modified the host file on my Windows, restarted the server and on the browser I've typed http://pwa.myapp.local/home, which should return me to the login page as I'm not connected, but I get this error from AdonisJS: `E_CANNOT_FIND_ROUTE: Cannot find route for ‘auth.user.login’ identifier`.
However, the route does appear on the `list:routes`!

If I go to http://pwa.myapp.local/login, the page is displayed...
In my doLogin controller (which is called when the form is sent) I've added a `console.log(‘ok’)`.
I enter my login, I validate, the ‘ok’ is displayed on the terminal, but I'm still located on /login (and not /home which should have appeared once the user had logged in) and I get this error: `E_CANNOT_FIND_ROUTE: Cannot find route for "front.home" identifier`, which again appear on the `list:routes`.

Have I forgotten to do something? Is adding the `domain` parameter to the route.ts not enough?
Why does it work by typing URLs directly, but calls from controllers (`return response.redirect().toRoute('front.home')`), Auth.ts (for redirect to login page if not logged) and templates (`{{ route('front.home') }}`) throw an error?

In case the problem was with nested groups, I tried another approach, but the result was the same:
```javascript
import Route from '@ioc:Adonis/Core/Route'

/* Login */
Route.get('/login', './Front/SecurityController.login').as('auth.user.login').domain('pwa.myapp.local') // Login
Route.post('/login', './Front/SecurityController.doLogin').domain('pwa.myapp.local')

/* App */
Route.group(() => {
  Route.get('/home', './Front/HomeController.index').as('home') // Home
  Route.get('/users/:slug', './Front/UsersController.show').as('users.show') // Users
})
  .as('front')
  .domain('pwa.myapp.local')
  .middleware('auth:user')
```

Does anyone know what the problem is? Thanks in advance and see you soon.

**Comment by RomainLanz** [comment, marked as answer]
Hey @BenDavidson90! 👋🏻 

Since you are using a domain, you have to specify the domain when generating the route.

```ts
Route
  .get('/users/:id', 'UsersController.show')
  .domain(':tenant.adonisjs.com')

// Make URL
const url = Route.makeUrl('UsersController.show', [1], {
  domain: ':tenant.adonisjs.com',
})
```

📚 https://v5-docs.adonisjs.com/guides/routing#lookup-inside-a-domain

**Reply by bsdev90** [reply]
I sorry for the big delay..
Thank you for your answer ! I really thought that just changing the route.ts was sufficient.
After your answer, thirst I thought that it will not fit my project because of domain not known in advance and set into the .env, but in fact there is always a solution !
In the route.ts:
```javascript
import Env from '@ioc:Adonis/Core/Env'

const domain = `${Env.get('ADMIN_SUBDOMAIN', 'admin')}.${Env.get('DOMAIN', 'localhost')}`

Route.group(() => {
...
}).domain(domain)
```
For the controllers I injected that in the request object request to obtain request.adminDomain and request.frontDomain, and for the templates I user a middleware to pass these infos.
The harder thing was to change all route call in the controllers and views, (more than 300 in my app), but that's my fault, I had to think of that before...
